app/src/wakeup_trigger_key.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2023 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/init.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/pm.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define DT_DRV_COMPAT zmk_wakeup_trigger_key
+
+struct wakeup_trigger_key_config {
+    struct gpio_dt_spec trigger;
+    size_t extra_gpios_count;
+    struct gpio_dt_spec extra_gpios[];
+};
+
+static int zmk_wakeup_trigger_key_init(const struct device *dev) {
+#if IS_ENABLED(CONFIG_PM_DEVICE)
+    pm_device_init_suspended(dev);
+    pm_device_wakeup_enable(dev, true);
+#endif
+
+    return 0;
+}
+
+#if IS_ENABLED(CONFIG_PM_DEVICE)
+
+static int wakeup_trigger_key_pm_action(const struct device *dev, enum pm_device_action action) {
+    const struct wakeup_trigger_key_config *config = dev->config;
+    int ret = 0;
+
+    switch (action) {
+    case PM_DEVICE_ACTION_RESUME:
+        ret = gpio_pin_interrupt_configure_dt(&config->trigger, GPIO_INT_LEVEL_ACTIVE);
+        if (ret < 0) {
+            LOG_ERR("Failed to configure wakeup trigger key GPIO pin interrupt (%d)", ret);
+            return ret;
```

Comment:
Minor: mixing early returns with allowing the case to run until the break and `return ret` at the end seems a bit weird.

Maybe pull each of these cases out into a separate function, so then you can just do this?

```c
switch (action) {
case PM_DEVICE_ACTION_RESUME:
    return wakeup_trigger_key_pm_resume(dev);

case PM_DEVICE_ACTION_SUSPEND:
    return wakeup_trigger_key_pm_suspend(dev);

default:
    return -ENOTSUP;
}
```

Reply:
Much better. Refactored.